### PR TITLE
replacement of pickle with json for suite database

### LINKED ISSUE #2567: suite run database: use JSON instead of pickle for timers data structures
Task action timers context and delays are currently serialized using pickle for storage in the suite run database. JSON has better support these days for serializing data structure and does not have some of the issues that pickle has.

We should replace pickle with JSON in this usage. The complication is that some of these data structures are Python named tuples, so they will need to be handled carefully on serializing and de-serializing.

### PR DESCRIPTION
**Pull request by lhuggett**
Closes #2567. Should be noted that the namedtuples being extracted from the database with json must all be listed specifically on retrieval.

### CODE REVIEW

**Comment by matthewrmshin** [comment]
This test `./tests/restart/20-event-retry.t` is failing in Travis CI and my environment. Found this traceback in the suite log.

```
Traceback (most recent call last):
  File "/home/matt/cylc.git/bin/cylc-restart", line 30, in <module>
    main(is_restart=True)
  File "/home/matt/cylc.git/lib/cylc/scheduler_cli.py", line 65, in main
    scheduler.start()
  File "/home/matt/cylc.git/lib/cylc/scheduler.py", line 231, in start
    self.run()
  File "/home/matt/cylc.git/lib/cylc/scheduler.py", line 1365, in run
    self.suite_db_mgr.put_task_event_timers(self.task_events_mgr)
  File "/home/matt/cylc.git/lib/cylc/suite_db_mgr.py", line 285, in put_task_event_timers
    ctx_obj[type(my_obj).__name__] = my_obj.__getnewargs__()
AttributeError: 'dict' object has no attribute '__getnewargs__'
```

**Comment by matthewrmshin** [comment]
Yes, we'll have a go at adding some upgrade logic.

**Comment by matthewrmshin** [comment]
Hi @lhuggett I have raised a PR against your branch to address the final upgrade and testing issues. See lhuggett/cylc#1.

**Comment by matthewrmshin** [comment]
(Given my involvement in this PR, @oliver-sanders please replace me as reviewer.)

**Comment by hjoliver** [comment]
Also, there's a dangling pickle import in `task_pool.py`.

**Comment by hjoliver** [comment]
Otherwise all good though - tests as workihng.

File: lib/cylc/scheduler.py [hunk 1]
```diff
@@ -214,7 +215,11 @@ def start(self):
         glbl_cfg().create_cylc_run_tree(self.suite)
 
         if self.is_restart:
-            self.suite_db_mgr.restart_upgrade()
+            try:
+                self.suite_db_mgr.restart_upgrade()
+            except OperationalError:
+                ERR.error("cannot upgrade task action timer table in database")
```
[the remark this answers is not in the document]
Comment (matthewrmshin) [reply lib/cylc/scheduler.py]: I am going to revert this change. A failed restart upgrade should print the traceback regardless.

File: lib/cylc/rundb.py [hunk 2]
```diff
@@ -904,6 +916,59 @@ def _upgrade_with_state_file_extras(self):
...
+                                {type(orig).__name__: orig.__getnewargs__()}))
+                        else:
+                            args.append(json.dumps(orig))
+                    elif col.endswith("_pickle"):
+                        # Upgrade pickled lists
+                        args.append(json.dumps(pickle.loads(str(cell))))
+                    else:
+                        args.append(cell)
+            except (EOFError, TypeError, LookupError, ValueError):
+                continue  # skip bad rows
```
[the remark this answers is not in the document]
Comment (matthewrmshin) [reply lib/cylc/rundb.py]: Will report number of skips.

**Comment by matthewrmshin** [comment]
Response to feedback @ lhuggett/cylc#2.

**Comment by matthewrmshin** [comment]
@hjoliver should now be good for a final look.
@sadielbartholomew please replace @oliver-sanders as reviewer while he is busy with docs.

**Comment by matthewrmshin** [comment]
(@lhuggett Sorry, I have taken another look at this one. Something has indeed gone a bit wrong as you suggested. I'll try my best to fix it again, and will probably supersede with a new pull request in due course. Individual changesets will still be attributed to you, so you will still be acknowledged as an author of the work. I hope you don't mind.)

**Comment by matthewrmshin** [comment]
Superseded by #2625.

**Pull request closed without merging**